OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/ide/instances/lcp_on_fields/problem/LinearConstantPropagationProblemExtended.scala

```diff
@@ -0,0 +1,280 @@
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation.problem.LinearConstantPropagationFact
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation.problem.LinearConstantPropagationProblem
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation.problem.LinearConstantPropagationValue
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation.problem.NullFact
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation.problem.UnknownValue
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation.problem.UnknownValueEdgeFunction
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation.problem.VariableFact
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation.problem.VariableValue
+import org.opalj.tac.fpcf.analyses.ide.instances.linear_constant_propagation.problem.VariableValueEdgeFunction
+import org.opalj.tac.fpcf.analyses.ide.solver.JavaStatement
+import org.opalj.tac.fpcf.analyses.ide.solver.JavaStatement.V
+
+/**
+ * Extended definition of the linear constant propagation problem, trying to resolve field accesses with the LCP on
+ * fields analysis.
+ *
+ * @author Robin Körkemeier
+ */
+class LinearConstantPropagationProblemExtended extends LinearConstantPropagationProblem {
+    override def isArrayLoadExpressionGeneratedByFact(
+        arrayLoadExpr: ArrayLoad[V]
+    )(
+        source:     JavaStatement,
+        sourceFact: LinearConstantPropagationFact,
+        target:     JavaStatement
+    ): Boolean = {
+        val arrayVar = arrayLoadExpr.arrayRef.asVar
+
+        /* Generate fact only if the variable represented by the source fact is one possible initializer of the index
+         * variable */
+        sourceFact match {
+            case NullFact =>
+                arrayVar.value.asReferenceValue.asReferenceType.asArrayType.componentType.isIntegerType
+
+            case VariableFact(_, definedAtIndex) =>
+                val indexVar = arrayLoadExpr.index.asVar
+                indexVar.definedBy.contains(definedAtIndex) &&
+                    arrayVar.value.asReferenceValue.asReferenceType.asArrayType.componentType.isIntegerType
+        }
+    }
+
+    override def getNormalEdgeFunctionForArrayLoad(
+        arrayLoadExpr: ArrayLoad[JavaStatement.V]
+    )(
+        source:     JavaStatement,
+        sourceFact: LinearConstantPropagationFact,
+        target:     JavaStatement,
+        targetFact: LinearConstantPropagationFact
+    )(implicit propertyStore: PropertyStore): EdgeFunctionResult[LinearConstantPropagationValue] = {
+        if (sourceFact == nullFact) {
+            return UnknownValueEdgeFunction
+        }
+
+        val arrayVar = arrayLoadExpr.arrayRef.asVar
+
+        val index = arrayLoadExpr.index.asVar.value match {
+            case intRange: DefaultIntegerRangeValues#IntegerRange =>
```

Comment:
Check if the type projection is necessary. OPAL is about to switch to Scala 3 which has dropped general type projection.